src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,170 @@
+                    value, findNodeWithIndex(index));
+            findNodeWithIndex(index).prev.next = node;
+            findNodeWithIndex(index).prev = node;
+            size++;
+            return;
+        }
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        isIndexGood(index);
+        if (index == size) {
+            throw new IndexOutOfBoundsException("Wrong index");
+        }
+        return findNodeWithIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        isIndexGood(index);
+        if (index == size) {
+            throw new IndexOutOfBoundsException("Wring index");
+        }
+        T valueBeforeSetting = findNodeWithIndex(index).item;
+        findNodeWithIndex(index).item = value;
+        return valueBeforeSetting;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        isIndexGood(index);
+        if (index == size) {
+            throw new IndexOutOfBoundsException("Wrong index");
+        }
+        Node<T> node = findNodeWithIndex(index);
+        T item = node.item;
+        if (index == size - 1 && index == 0) {
+            first.item = null;
+            last.item = null;
+            size--;
+        } else if (index == 0 && index != size - 1) {
+            node.next.prev = null;
+            first = node.next;
+            node.next = null;
+            node.item = null;
+            size--;
+        } else if (index == size - 1 && index != 0) {
+            node.prev.next = null;
+            last = node.prev;
+            node.prev = null;
+            node.item = null;
+            size--;
+        } else {
+            node.prev.next = node.next;
+            node.next.prev = node.prev;
+            unlinkNode(node);
+            size--;
+        }
+        return item;
     }
 
     @Override
     public boolean remove(T t) {
+        Node<T> node = first;
+        for (int i = 0; i < size; i++) {
+            if (node.item == t
+                    || (node.item != null && node.item.equals(t))) {
+                remove(i);
+                return true;
+            }
+            node = node.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    public boolean isIndexGood(int index) {
+        if (index > size || index < 0) {
+            throw new IndexOutOfBoundsException("Wrong index");
+        } else {
+            return true;
+        }
+    }
+
+    public Node<T> findNodeWithIndex(int index) {
+        Node<T> nodeWithIndex = first;
```

Comment:
in this method let's start moving from the first Entry OR from the last one
depending on whether index is closer to the start or to the end’